### DISCUSSION TITLE
EXE export with Pointer Lock requires initial interaction from mousebutton

### DISCUSSION THREAD

**Discussion by entangledothers** [cables-gl/cables_docs#338] [! Welcome - readme first!]
Hi cables team!
So I've stumbled over a small headache, I have a 3D `AmmoWorld` scene setup with an `AmmoCharacterFpsCamera` who has the  ` Enable Pointer Lock` set to `True`, all this works beautifully in browser, however for an exhibition I need to export this as an exe as it will be shown inside a faux game cabinet with joysticks as inputs. 

The executable exports just fine, but when the app is first opened the Pointer Lock function doesn't grab the cursor, so until a mouse is used to click on the window, any mouse/joystick input will not work until a mouse has clicked on the window. This behaviour is consistent on windows and linux (haven't tested mac yet, but assume it's the same). Building an option for adding/removing a mouse to the hardware apparently isn't possible at this late stage, nor can the patch be left running 24/7 in the name of energy saving.

Is there a way to fix this issue so that for executable exports pointer lock "just works" without requiring a mouse?

**Comment by steam0r** [comment]
hey,

we reworked the exe export (to not need a github account anymore) and also fixed some other minor issues. we think this might be fixed in the new version already. can you try exporting your patch to exe on dev.cables.gl and report back (and also any other issues that you might find with the new export). 

thanks!

**Reply by entangledothers** [reply]
Thanks for the prompt reply, I've tested the EXE export on dev.cables.gl, but the issue persists. Only after a mouseclick does the pointer lock work as expected.

**Reply by steam0r** [reply]
can you maybe provide a simple example patch? 

it seems that the "physics room" template allows for immediate interaction in the exe export. 

pointerlock will never work without clicking the mouse, as it needs an interaction to work. am i getting this issue wrong?

**Reply by entangledothers** [reply]
Gotcha, as a mouse click is an immutable requirement then I'll have to find an alternative approach. Thanks for clarifying!

**Reply by entangledothers** [reply]
[Here's an example patch](https://dev.cables.gl/p/631ef8154239227b62351041?s=lJYt0lek3Atmjpmw). When exported as an EXE it will always need a mouse click before any mouse/similar input will work to manipulate the camera, this naturally has consequences for intuitive use and installation setups where no mouse is used.

**Reply by steam0r** [reply]
you mean WASD does not work for you in that example patch when exported to exe?

**Reply by entangledothers** [reply]
Not until the mouse has been clicked in order to "focus" the window: all input is ignored until that is done.

**Comment by pandrr** [comment]
ok i added a trigger input to canvasFocus on dev.cables.gl, you can trigger this once in the beginning and then the canvas is properly focussed. 


![Screenshot 2022-09-12 at 20 01 43](https://user-images.githubusercontent.com/98792/189724425-6cdad5b2-5a88-4a18-a357-c14a185f099d.png)

**Reply by entangledothers** [reply]
This fixed keyboard input, now WASD/any other keys work without having to click the mouse first. Is there any chance this could also be tweaked to include the mouse input?

**Comment by pandrr** [comment]
that op expects the mouse to be pressed

**Reply by entangledothers** [reply]
OK, since there seems to be no alternative op for that I'll have to try wrapping inside a script that triggers a simulated mouse click on initial start. Thanks for fixing the keyboard inputs!

**Reply by pandrr** [reply]
you cant fake this, clone the op and remove the check if the button is down, its easy

**Reply by entangledothers** [reply]
Well, perhaps not with javascript, but an external script wrapper that executes the patch and handles the mouseclick event works just fine. Terribly hacky, but saves me redoing this in unity for now.
